pkg/controller/wip/controller.go

```diff
@@ -119,8 +125,151 @@ func (c *controller) brokerDelete(obj interface{}) {
+	} else {
+		glog.V(5).Infof("Updated ready condition for Broker %v to True", broker.Name)
+	}
+}
+
+// reconcileServiceClassFromBrokerCatalog reconciles a ServiceClass after the
+// Broker's catalog has been re-listed.
+func (c *controller) reconcileServiceClassFromBrokerCatalog(broker *v1alpha1.Broker, serviceClass *v1alpha1.ServiceClass) error {
+	serviceClass.BrokerName = broker.Name
+
+	existingServiceClass, err := c.serviceClassLister.Get(serviceClass.Name)
+	if err != nil {
+		// An error returned from a lister Get call means that the object does
+		// not exist.  Create a new ServiceClass.
+		if _, err := c.serviceCatalogClient.ServiceClasses().Create(serviceClass); err != nil {
+			glog.Errorf("Error creating serviceClass %v from Broker %v: %v", serviceClass.Name, broker.Name, err)
+			return err
+		}
+	}
+
+	clone, err := api.Scheme.DeepCopy(existingServiceClass)
+	if err != nil {
+		return err
+	}
+
+	toUpdate := clone.(*v1alpha1.ServiceClass)
+	toUpdate.Bindable = serviceClass.Bindable
+	toUpdate.Plans = serviceClass.Plans
+	toUpdate.PlanUpdatable = serviceClass.PlanUpdatable
+	toUpdate.OSBTags = serviceClass.OSBTags
+	toUpdate.OSBRequires = serviceClass.OSBRequires
+	toUpdate.OSBMaxDBPerNode = serviceClass.OSBMaxDBPerNode
+	toUpdate.OSBDashboardOAuth2ClientID = serviceClass.OSBDashboardOAuth2ClientID
+	toUpdate.OSBDashboardSecret = serviceClass.OSBDashboardSecret
+	toUpdate.OSBDashboardRedirectURI = serviceClass.OSBDashboardRedirectURI
+
+	toUpdate.Description = serviceClass.Description
+	toUpdate.DisplayName = serviceClass.DisplayName
+	toUpdate.ImageURL = serviceClass.ImageURL
+	toUpdate.LongDescription = serviceClass.LongDescription
+	toUpdate.ProviderDisplayName = serviceClass.ProviderDisplayName
+	toUpdate.DocumentationURL = serviceClass.DocumentationURL
+	toUpdate.SupportURL = serviceClass.SupportURL
+
+	if _, err := c.serviceCatalogClient.ServiceClasses().Update(toUpdate); err != nil {
+		glog.Errorf("Error updating serviceClass %v from Broker %v: %v", serviceClass.Name, broker.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+func (c *controller) updateBrokerReadyCondition(broker *v1alpha1.Broker, status v1alpha1.ConditionStatus, reason, message string) error {
+	var (
+		foundReadyCondition = false
+		readyConditionIndex = 0
+		newCondition        = v1alpha1.BrokerCondition{
+			Type:    v1alpha1.BrokerConditionReady,
+			Status:  status,
+			Reason:  reason,
+			Message: message,
+		}
+	)
+
+	for ii, condition := range broker.Status.Conditions {
+		if condition.Type == v1alpha1.BrokerConditionReady {
+			foundReadyCondition = true
+			readyConditionIndex = ii
+		}
+	}
+
+	clone, err := api.Scheme.DeepCopy(broker)
+	if err != nil {
+		return err
+	}
+	toUpdate := clone.(*v1alpha1.Broker)
+
+	if foundReadyCondition {
+		toUpdate.Status.Conditions[readyConditionIndex] = newCondition
+	} else {
+		toUpdate.Status.Conditions = append(toUpdate.Status.Conditions, newCondition)
+	}
+
```

Comment:
Not sure I understand most of the above. Why are we branching on whether a readycondition was already found in the ledger or statuses? I think there's also room for a big nasty bug here. Your loop breaks on the _first_ readycondition found and overwrites it. Is that right??? What if there's a more recent readycondition? I just don't get what's going on here.